RFC: rename if not exists (#7818)
apache/opendal: core/core/src/docs/rfcs/7818_rename_if_not_exists.md

- Proposal Name: `rename_if_not_exists`
- Start Date: 2026-06-24
- RFC PR: [apache/opendal#7818](https://github.com/apache/opendal/pull/7818)

# Summary

Extend rename with an `if_not_exists` option:

```rust
op.rename_with("staging/file", "published/file")
    .if_not_exists(true)
    .await?;
```

The existing `rename` API keeps its overwrite semantics. When
`if_not_exists` is enabled, rename succeeds only if the destination does not
exist. A destination conflict returns `ConditionNotMatch` without modifying the
source or destination.

# Motivation

OpenDAL defines `rename` as an overwrite operation. Some applications also need
an atomic publish primitive: move a completed staging file into place only when
no other writer has already published that destination.

A caller cannot implement this safely with `stat` followed by `rename`. Another
writer can create the destination after `stat` reports that it is absent but
before rename runs. A service configuration flag is also unsuitable because it
would make the meaning of the same `rename` call depend on backend construction

> rather than an explicit call-site condition.

Reviewer: ````suggestion
`stat` followed by `rename` are two atomic actions.
A user might want to move a file to destination in an atomic action.

An example is that when user is trying to:

```
stat_result = stat(destination)
if !stat_result {
  rename(source, destination)
}
```

Another writer could create destination file during the critical section.

A service configuration flag is also unsuitable because it
would make the meaning of the same `rename` call depend on backend construction
rather than an explicit call-site condition.
````

OpenDAL already models the equivalent destination condition for write and copy

> through options:

Reviewer: ```suggestion
We will follow how OpenDAL models `write` and `copy` with options:
```

No need to have "Rename should follow the same public API and error model." after that

```rust
op.write_with("path", content)
    .if_not_exists(true)
    .await?;

op.copy_with("source", "target")
    .if_not_exists(true)
    .await?;
```

Rename should follow the same public API and error model.

# Guide-level explanation

> Use `rename` when the destination may be replaced:

Reviewer: ```suggestion
Current behavior: use `rename` when users want to overwrite a destination file when the destination file exist:
```

```rust
use opendal::{Operator, Result};

async fn replace(op: Operator) -> Result<()> {
    op.rename("staging/file", "published/file").await?;
    Ok(())
}
```

Use `rename_with(...).if_not_exists(true)` when an existing destination must be

> preserved:

Reviewer: ```suggestion
New API: use `rename_with(...).if_not_exists(true)` when an existing destination must be
preserved when the destination file exist:
```

```rust
use opendal::{ErrorKind, Operator, Result};

async fn publish(op: Operator) -> Result<()> {
    match op
        .rename_with("staging/file", "published/file")
        .if_not_exists(true)
        .await
    {
        Ok(()) => Ok(()),
        Err(err) if err.kind() == ErrorKind::ConditionNotMatch => Err(err),
        Err(err) => Err(err),
    }
}
```

The conditional operation has the following outcomes:

- If the destination does not exist, the source is renamed to the destination.
- If the destination exists, the operation returns `ConditionNotMatch` and
  leaves both paths unchanged.
- If the service cannot enforce the destination condition atomically, the

>   operation returns `Unsupported`.

Reviewer: Thorough thought to have this exception.

Since we have a capability flag, should we allow users to construct options to call a service?

I didn't look into the `copy` or `write` code. But I would see values aligning a similar behavior.

Author: @erickguan Thanks！
This aligns with how copy and write handle the same situation. The options struct is always constructable — there's no compile-time gate. At runtime, the `correctness_check` layer inspects the capability flag and returns `Unsupported` when `if_not_exists` is set but the service doesn't advertise it:

- write: [correctness_check.rs#L136-L143](https://github.com/apache/opendal/blob/main/core/core/src/layers/correctness_check.rs#L136-L143)
- copy:  [correctness_check.rs#L210-L217](https://github.com/apache/opendal/blob/main/core/core/src/layers/correctness_check.rs#L210-L217)

Rename follows the identical pattern. I've added a note in the RFC explicitly calling out this alignment.

```rust
pub struct OpRename {
    if_not_exists: bool,
}
```

The `Service::rename` signature does not change. Services inspect
`OpRename::if_not_exists()` to select their native overwrite or no-overwrite
operation.

Add a capability field:

```rust
pub struct Capability {
    pub rename: bool,
    pub rename_with_if_not_exists: bool,
}
```

`rename_with_if_not_exists` is meaningful only when `rename` is also supported.
The correctness check returns `Unsupported` before dispatch when the option is
enabled but the service does not advertise the capability.

## Error semantics

`if_not_exists` is a destination precondition, so a destination conflict maps
to `ConditionNotMatch`. This is consistent with existing write and copy
behavior.

Backend errors caused by the native no-replace condition must be translated to
`ConditionNotMatch`, even when the backend or operating system reports a native
error such as `AlreadyExists`.

Other errors retain their normal meanings. For example, a missing source
returns `NotFound`, a directory passed where a file is required returns the
corresponding directory error, and an unsupported condition returns
`Unsupported`.

> ## Atomicity requirement

Reviewer: Nice! Clear descriptions

Author: Thanks!

A service must advertise `rename_with_if_not_exists` only when the destination
condition and rename are enforced by one native operation. Performing a
destination check followed by an overwriting rename has a time-of-check to
time-of-use race and does not satisfy the capability.

> When the destination condition fails:

Reviewer: What is "the destination condition"?

Author: Thanks for catching that — replaced 'the destination condition' with explicit phrasing: 'destination-does-not-exist check' and 'the destination already exists (the if_not_exists condition fails)'.

- The source remains at its original path.
- The destination remains unchanged.

> When the operation succeeds:

Reviewer: What is "the operation"?

Author: Replaced with 'the rename' to avoid the ambiguity. In this section each occurrence now refers to an explicit term: the rename, the check, the native operation.

Author: have fixed

- The source no longer exists at its original path.
- The destination contains the source data.

This RFC does not require rename to be crash-atomic beyond the guarantees of the
underlying service. It requires the no-replace decision to be atomic with
respect to competing destination creation.

## Service analysis

### Audit scope

At the time of this RFC, 19 services advertise `Capability::rename`. The table
below audits all of them, plus S3 as the representative object store without a
native rename operation.

The "assessment" column describes whether the current backend primitive could
support `rename_with_if_not_exists`. It does not enable the capability. A
backend must still demonstrate that the destination check and move are one
atomic service operation and map destination conflicts to
`ConditionNotMatch`.

| Service | Current rename implementation | Assessment for `if_not_exists` |
| --- | --- | --- |

> | `aliyun_drive` | Deletes the destination, moves by file ID, then updates the name | Not suitable as implemented; overwrite is a multi-request workflow |

Reviewer: This current implementation contradicts atomicity you are introducing. Should `rename` with overwrite also be atomic?

Author: Good catch. The key distinction: overwrite rename does NOT need atomicity because its contract is simply "the destination ends up with the source content." A multi-step delete+rename achieves this. But if_not_exists IS conditional — it must check "destination absent" AND move atomically, or a concurrent writer slips in. I have added a paragraph before the service analysis section explaining that a backend with non-atomic overwrite may still support conditional rename when its native move rejects conflicts atomically. HDFS is the concrete example.

Any existing cross-platform differences in default FS overwrite behavior are
outside this RFC and should be handled separately.

`compfs` and `monoiofs` wrap the same operating-system rename model through
different runtimes, so they have the same portability constraint. SFTP also
requires caution because the effective rename behavior depends on the protocol
operation and server implementation.

### Native conditional move services

Several services expose a destination policy in the same server-side move
request:

- Azure Data Lake Storage accepts `If-None-Match: *` for rename.
- Azure Files exposes `x-ms-file-rename-replace-if-exists`.
- WebDAV defines the `Overwrite` header for `MOVE`; `F` requires the server to
  reject an existing destination.
- Yandex Disk exposes an `overwrite` query parameter.
- OneDrive exposes conflict behavior, although support differs between service
  variants and must be verified before advertising the capability.

These services are candidates for follow-up implementations. They are not
enabled by this RFC's initial HDFS change. Each follow-up must include behavior
coverage and service-specific error translation.

### Multi-step overwrite services

> `aliyun_drive`, `gdrive`, `goosefs`, and `koofr` currently delete or trash an

Reviewer: Depending on if we are changing `rename` to be atomic, say yes. Then we are essentially making these services' rename to be a simulation. Either this is way beyond the scope and we register this problem as a future work. Or we would see a way to address this problem in this RFC. e.g., a capability flag for declaring so? or documentation on atomicity with/without `rename_with_if_not_exists`.

Author: I think the answer is neither — this RFC does NOT propose making rename atomic. Multi-step delete+rename is a valid implementation of overwrite, not a "simulation." It genuinely fulfills the contract: the destination ends up with the source content.

The atomicity gate is introduced only for rename_with_if_not_exists because it adds a conditional check that would be meaningless without atomic enforcement. Overwrite rename (the default) has no such check, so no new atomicity requirement.

That said, a capability flag like rename_is_atomic for the base rename operation is a reasonable idea, but it is separate from this RFC. I would prefer keeping this RFC scoped to if_not_exists and filing a follow-up issue for rename atomicity documentation/capability if the maintainers agree it is worth tracking.

Reviewer: @hfutatzhanghb  Is zhb your handle on OpenDAL discord?
